### PR TITLE
feat(gke-cluster-standard): Add optional `boot_disk_kms_key` into node_config

### PR DESCRIPTION
**Pull request by anthonyhaussman**
Add optional setup to add `boot_disk_kms_key` into the `node_config`.

Without it, configuring a standard GKE cluster with the policy `constraints/gcp.restrictNonCmekServices` returns an error:
```
│ Error: googleapi: Error 400: Failed precondition: Constraint `constraints/gcp.restrictNonCmekServices` violated for `projects/XXXXXXXX` attempting to create a resource without specifying a KMS CryptoKey.
│ Details:
│ [
│   {
│     "@type": "type.googleapis.com/google.rpc.RequestInfo",
│     "requestId": "0xde962b1494496ff1"
│   }
│ ]
│ , failedPrecondition
```

---
**Checklist**

I applicable, I acknowledge that I have:
- [X] Read the [contributing guide](https://github.com/GoogleCloudPlatform/cloud-foundation-fabric/blob/master/CONTRIBUTING.md)
- [X] Ran `terraform fmt` on all modified files
- [X] Regenerated the relevant README.md files using [`tools/tfdoc.py`](https://github.com/GoogleCloudPlatform/cloud-foundation-fabric/blob/master/CONTRIBUTING.md#fabric-tools)
- [x] Made sure all relevant tests pass

### CODE REVIEW

File: modules/gke-cluster-standard/main.tf [hunk 1]
```diff
@@ -128,7 +129,7 @@ resource "google_container_cluster" "cluster" {
       dynamic "auto_provisioning_defaults" {
         for_each = var.cluster_autoscaling.auto_provisioning_defaults != null ? [""] : []
         content {
-          boot_disk_kms_key = var.cluster_autoscaling.auto_provisioning_defaults.boot_disk_kms_key
+          boot_disk_kms_key = var.boot_disk_kms_key != null ? var.boot_disk_kms_key : var.cluster_autoscaling.auto_provisioning_defaults.boot_disk_kms_key
```
Comment (ludoo) [inline modules/gke-cluster-standard/main.tf:132]: why don't you use the existing field in auto provisioning defaults?

Comment (anthonyhaussman) [reply modules/gke-cluster-standard/main.tf]: As `boot_disk_kms_key` is a "root" variable, it makes sense to set the default KMS to the "root" variable if it's available.
I have changed the logic in my last commit to using in `auto_provisioning_defaults` first the variable ` var.cluster_autoscaling.auto_provisioning_defaults.boot_disk_kms_key` if it exists and fallback to `var.boot_disk_kms_key` if it exists.

What do you think?

Comment (ludoo) [reply modules/gke-cluster-standard/main.tf]: I don't think we need two variables for the same thing. :) And in general we always try to keep our variable spaces small. I am closing this, but thanks anyway.

**Comment by anthonyhaussman** [comment]
@ludoo,
Can I know why did you close the PR? :confused: 
Is that possible to have an explanation for this? This is a blocker on our side to push strict Org policies.

File: modules/gke-cluster-standard/main.tf (hunk 1 above)
Comment (anthonyhaussman) [reply modules/gke-cluster-standard/main.tf]: But without the setting into the `node_config` configuration, the constraint policy `constraints/gcp.restrictNonCmekServices` blocks everything and the module can't be used at all.

Do we have a solution?

**Comment by ludoo** [comment]
Again: why don't you use the existing variable for the same thing?

**Comment by ludoo** [comment]
BTW you're free to fork and do your own changes, of course.

**Comment by ludoo** [comment]
Had a chat with @wiktorn and while this is indeed a problem, the proposed solution does not fit into our design patterns (and I did an awful job at explaining it, sorry for that): adding a top-level variable for a single value is a "waste of variable space" which we always try to avoid.

The right approach is to add a top-level variable for all the current `node_config` attributes (service account, tags, etc.) and add the KMS key there. Then using that in autoscale config with a condition much like in your PR, probably.

So we can now either

- update this PR to do the above, if you still feel like it, and also check that the autopilot module has the same changes
- do it ourselves

In the meantime, we'll turn this into an issue for tracking. Thanks for your patience, and for sticking with it. :)

**Comment by anthonyhaussman** [comment]
> Had a chat with @wiktorn and while this is indeed a problem, the proposed solution does not fit into our design patterns (and I did an awful job at explaining it, sorry for that): adding a top-level variable for a single value is a "waste of variable space" which we always try to avoid.
> 
> The right approach is to add a top-level variable for all the current `node_config` attributes (service account, tags, etc.) and add the KMS key there. Then using that in autoscale config with a condition much like in your PR, probably.
> 
> So we can now either
> 
> * update this PR to do the above, if you still feel like it, and also check that the autopilot module has the same changes
> * do it ourselves
> 
> In the meantime, we'll turn this into an issue for tracking. Thanks for your patience, and for sticking with it. :)

Gonna try to propose something in this PR. 
Thanks for your explanations. :pray:

**Comment by ludoo** [comment]
Thanks for flagging this! PR is incoming, I had a free half hour. :)

**Comment by anthonyhaussman** [comment]
> Thanks for flagging this! PR is incoming, I had a free half hour. :)

Thanks a lot for the reactivity :pray: 
It helps a lot! Many thanks! :pray:

**Comment by ludoo** [comment]
Thank you for being patient despite my pushback! :)

**Pull request closed without merging**